ssl/ssl_key_share.cc

```diff
@@ -294,6 +295,233 @@ class CECPQ2KeyShare : public SSLKeyShare {
   HRSS_private_key hrss_private_key_;
 };
 
+class KEMKeyShare : public SSLKeyShare {
+ public:
+  KEMKeyShare(int nid, uint16_t group_id) : nid_(nid), group_id_(group_id) {}
+
+  uint16_t GroupID() const override { return group_id_; }
+
+  bool Offer(CBB *out) override {
+    // Ensure |out| is valid, has no children, and has been initialized.
+    // We check that |out| has no children because otherwise CBB_data()
+    // will produce a fatal error by way of assert(cbb->child == NULL).
+    if (!out || out->child || !CBB_data(out)) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_PASSED_NULL_PARAMETER);
+      return false;
+    }
+
+    // ctx_ should not have been initialized (by e.g. a previous call
+    // to Offer()).
+    if (ctx_) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
+      return false;
+    }
+
+    // Initialize the KEM ctx
+    ctx_.reset(EVP_PKEY_CTX_new_id(EVP_PKEY_KEM, nullptr));
+    if (!ctx_ || !EVP_PKEY_CTX_kem_set_params(ctx_.get(), nid_)) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+      return false;
+    }
+
+    // Generate a KEM keypair and retrieve the length of the public key
+    EVP_PKEY *raw_key = nullptr;
+    if (!EVP_PKEY_keygen_init(ctx_.get()) ||
+        !EVP_PKEY_keygen(ctx_.get(), &raw_key) ||
+        !raw_key ||
+        !EVP_PKEY_get_raw_public_key(raw_key, nullptr, &public_key_len_)) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
```

Comment:
`raw_key` is not freed in the failure path.

Reply:
Fixed

Reply:
You're right, the unit tests I'm adding do not get into this failure case. I also don't see a way to trigger it; this failure path only triggers if one of the EVP APIs fails. I'm not sure how it would be possible to force one of those functions to fail from the outside like this. I'll give it some thought and check the EVP/KEM unit tests; maybe it's done there somehow in a way we can replicate here.